File: github-mcp-server/src/tools/commits.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { githubRequest, handleGitHubError } from "../services/github-client.js";
import { DEFAULT_PER_PAGE, MAX_PER_PAGE } from "../constants.js";
import type { GitHubCommit } from "../types.js";

export function registerCommitTools(server: McpServer): void {
  server.registerTool(
    "github_list_commits",
    {
      title: "List GitHub Commits",
      description: `List commits on a branch (or the default branch if none given), most recent first.

Args:
  - owner (string), repo (string)
  - sha (string, optional): Branch name, tag, or commit SHA to start listing from. Defaults to the default branch.
  - per_page (number, default 30, max 100)
  - page (number, default 1)

Returns JSON with a "commits" array (sha, message [first line], author_login, author_name, date, html_url) and pagination info.`,
      inputSchema: {
        owner: z.string().min(1),
        repo: z.string().min(1),
        sha: z.string().optional().describe("Branch, tag, or commit SHA to list from"),
        per_page: z.number().int().min(1).max(MAX_PER_PAGE).default(DEFAULT_PER_PAGE),
        page: z.number().int().min(1).default(1),
      },
      annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: true },
    },
    async ({ owner, repo, sha, per_page, page }) => {
      try {
        const data = await githubRequest<GitHubCommit[]>(`/repos/${owner}/${repo}/commits`, "GET", undefined, {
          sha,
          per_page,
          page,
        });
        const output = {
          page,
          per_page,
          count: data.length,
          has_more: data.length === per_page,
          commits: data.map((c) => ({
            sha: c.sha,
            message: c.commit.message.split("\n")[0],
            author_login: c.author?.login ?? null,
            author_name: c.commit.author?.name ?? null,
            date: c.commit.author?.date ?? null,
            html_url: c.html_url,
          })),
        };
        return {
          content: [{ type: "text", text: JSON.stringify(output, null, 2) }],
          structuredContent: output,
        };
      } catch (error) {
        return { content: [{ type: "text", text: handleGitHubError(error, "github_list_commits") }] };
      }
    }
  );
}

File: github-mcp-server-python/src/github_mcp/tools/commits.py
"""Commit history read tools."""

import json
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from ..app import mcp
from ..client import github_request, handle_github_error
from ..constants import DEFAULT_PER_PAGE, MAX_PER_PAGE


class ListCommitsInput(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True, extra="forbid")

    owner: str = Field(..., min_length=1)
    repo: str = Field(..., min_length=1)
    sha: Optional[str] = Field(default=None, description="Branch, tag, or commit SHA to list from")
    per_page: int = Field(default=DEFAULT_PER_PAGE, ge=1, le=MAX_PER_PAGE)
    page: int = Field(default=1, ge=1)


@mcp.tool(
    name="github_list_commits",
    annotations={
        "title": "List GitHub Commits",
        "readOnlyHint": True,
        "destructiveHint": False,
        "idempotentHint": True,
        "openWorldHint": True,
    },
)
async def github_list_commits(params: ListCommitsInput) -> str:
    """List commits on a branch (or the default branch if none given), most
    recent first.

    Args:
        params (ListCommitsInput): owner, repo, sha (optional branch/tag/commit
            to start from), per_page (default 30, max 100), page (default 1)

    Returns:
        str: JSON with "commits" array (sha, message [first line],
        author_login, author_name, date, html_url) and pagination info.
    """
    try:
        data = await github_request(
            f"/repos/{params.owner}/{params.repo}/commits",
            params={"sha": params.sha, "per_page": params.per_page, "page": params.page},
        )
        commits = [
            {
                "sha": c["sha"],
                "message": c["commit"]["message"].split("\n", 1)[0],
                "author_login": (c.get("author") or {}).get("login"),
                "author_name": (c["commit"].get("author") or {}).get("name"),
                "date": (c["commit"].get("author") or {}).get("date"),
                "html_url": c["html_url"],
            }
            for c in data
        ]
        output = {
            "page": params.page,
            "per_page": params.per_page,
            "count": len(commits),
            "has_more": len(commits) == params.per_page,
            "commits": commits,
        }
        return json.dumps(output, indent=2)
    except Exception as e:
        return handle_github_error(e, "github_list_commits")
